### PR TITLE
C++ feature tests

### PR DESCRIPTION
**Pull request by K-ballo**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -639,24 +639,126 @@ hpx_add_compile_flag(-DBOOST_DISABLE_ASSERTS
...
+
+hpx_check_for_cxx11_deleted_functions(
+  DEFINITIONS HPX_HAVE_CXX11_DELETED_FUNCTIONS)
+
+hpx_check_for_cxx11_explicit_cvt_ops(
+  DEFINITIONS HPX_HAVE_CXX11_EXPLICIT_CONVERSION_OPERATORS)
+
+hpx_check_for_cxx11_function_template_default_args(
+  REQUIRED "HPX needs support for C++11 defaulted function template arguments")
+
```
Comment (hkaiser) [inline CMakeLists.txt]: Where do we rely on defaulted function template arguments? I thought we wouldn't...

File: CMakeLists.txt [hunk 2]
```diff
@@ -639,24 +639,126 @@ hpx_add_compile_flag(-DBOOST_DISABLE_ASSERTS
...
+hpx_check_for_cxx11_std_type_traits(
+  DEFINITIONS HPX_HAVE_CXX11_STD_TYPE_TRAITS)
+
+# Check the availability of certain C++14 language features
+hpx_check_for_cxx14_constexpr(
+  DEFINITIONS HPX_HAVE_CXX14_CONSTEXPR)
+
+hpx_check_for_cxx14_lambdas(
+  DEFINITIONS HPX_HAVE_CXX14_LAMBDAS)
+
```
Comment (hkaiser) [inline CMakeLists.txt]: Should we add a test for C++11 lambdas? We rely on those, now.

**Comment by hkaiser** [comment]
Otherwise, LGTM! Thanks!

File: CMakeLists.txt (hunk 1 above)
Comment (K-ballo) [inline CMakeLists.txt]: Variadic `tuple` relies on it, as it needs to do SFINAE on a variadic constructor.

File: CMakeLists.txt (hunk 2 above)
Comment (K-ballo) [inline CMakeLists.txt]: There's one up here, in the C++11 section

**Pull request merged**